// SVC/financeSvc.js
/**
 * 财务服务
 */
var Svc = require('./BaseSvc').BaseSvc;
var db = Svc.db;
var async = require('async');
var _ = require('underscore');
var FinanceSvc = exports.FinanceSvc = {
    /**
     * 创建应收应付
     * @param {Object} owner 所有者{Name:String ,Value:String}
     * @param {Object} relativeOrg 相关组织 {Name:String,Value:String}
     * @param {String} summary 摘要
     * @param {Number} debit 借方j金额
     * @param {Number} credit 贷方金额
     * @param {Object} creator 创建人
     * @param {String} remark 备注
     * @param cb 回调 参数(error,RnP实例)
     */
    CreateRnP: function (owner, relativeOrg, summary, debit, credit, creator, remark, cb) {
        Svc.db.RnP.find({'Org.Value': owner.Value, 'RelativeOrg.Value': relativeOrg.Value}).sort({'CreateTime.Item1': -1}).limit(1).toArray(function (e, d) {
            var rm = d.length == 0 ? 0 : d[0].Remain;
            Svc.db.RnP.insert({
                _id:Svc.db.RnP.ObjectID().toString(),
                Creator: creator,
                CreateTime: Date.ToCreateTime(),
                Org: owner,
                RelativeOrg: relativeOrg,
                Summary: summary,
                Debit: debit,
                Credit: credit,
                Remain: rm + debit - credit,
                Remark: remark,
                flag: 1
            }, function (e, ds) { cb(e, ds[0]);});
        });
    }
};
